src/commonMain/kotlin/character/races/abilities/DevouringPlague.kt

```diff
@@ -0,0 +1,75 @@
+package character.races.abilities
+
+import character.Ability
+import character.Buff
+import character.Proc
+import character.Stats
+import character.classes.priest.Priest
+import character.classes.priest.buffs.InnerFocus as InnerFocusBuff
+import character.classes.priest.talents.ShadowFocus
+import character.races.Undead
+import character.races.debuffs.DevouringPlagueDot
+import data.Constants
+import mechanics.General
+import mechanics.Spell
+import sim.Event
+import sim.EventResult
+import sim.EventType
+import sim.SimParticipant
+
+class DevouringPlague : Ability() {
+    override val id: Int = 19280
+    override val name: String = "Devouring Plague"
+
+    val school = Constants.DamageType.SHADOW
+    val baseDamage = 1216.0
+    val baseDotTickCount = 8
+    // https://wowwiki-archive.fandom.com/wiki/Spell_power_coefficient?oldid=1492745
+    val spellPowerCoeff = 0.8
+
+    override fun cooldownMs(sp: SimParticipant): Int = 180000
+
+    override fun gcdMs(sp: SimParticipant): Int = sp.spellGcd().toInt()
+
+    override fun available(sp: SimParticipant): Boolean {
+        return sp.character.race is Undead && sp.character.klass is Priest && super.available(sp)
+    }
+
+    val baseResourceCost = 1145.0
+    override fun resourceCost(sp: SimParticipant): Double {
+        val innerFocusBuff = sp.buffs[InnerFocusBuff.name] as InnerFocusBuff?
+
+        if (innerFocusBuff != null) {
+            return 0.0
+        }
+
+        return baseResourceCost
+    }
+
+    override fun cast(sp: SimParticipant) {
+        val sfTalent: ShadowFocus? = sp.character.klass.talentInstance(ShadowFocus.name)
+        val sfHit = sfTalent?.shadowHitIncreasePct() ?: 0.0
+
+        // snapshot damage on initial cast
+        val damageRoll = Spell.baseDamageRollSingle(sp, baseDamage, school, spellPowerCoeff)
+        val result = Spell.attackRoll(sp, damageRoll, school, isBinary = true, bonusHitChance = sfHit, canCrit = false)
+
+        val event = Event(
```

Comment:
I had copied the initial cast events from SiphonLife / Corruption for the dots I implemented before I realized this was done automatically. For the purposes of dot initial application events with Spell Hits / Resists, what should be the correct event type here? I could use `EventType.Damage`, but it feels weird arbitrarily assigning that event when no damage is being done for initial dot application. I think we need a way to grab the spell cast event for passing to `SPELL_HIT` / `SPELL_RESIST` procs in this case.

Reply:
So something like below in `SimParticipant`
```
// Log cast event
  val castEvent = Event(
      eventType = EventType.SPELL_CAST,
      abilityName = castingRule!!.ability.name,
      target = sim.target
  )

  castingRule!!.ability.beforeCast(this)
  castingRule!!.ability.cast(this, castEvent)
  castingRule!!.ability.afterCast(this)

  logEvent(castEvent)
```
and update the `Ability` to accept an optional cast event:
```
abstract fun cast(sp: SimParticipant, castEvent: Event?)
```

Reply:
This would actually clean up quite a few spells since we don't need to generate a no-op event for triggering spell based procs, but what ever you think works here.

Reply:
> For the purposes of dot initial application events with Spell Hits / Resists, what should be the correct event type here?

In order to work with sim stats output, it would currently have to be a `DAMAGE` event - those all get rolled up into the main abilities table.  I'm not sure how useful it would be to have a new event, since it would have to have all the fields a `DAMAGE` event would normally have, anyway, so the stats code can deal with it.

> So something like below in SimParticipant

I'm not super excited about intentionally mutating an argument, or about changing the signature of `cast` - that would require some rework to every ability, due to how Kotlin handles override signatures.  

I think adding the zero `DAMAGE` event is best, since it will fit most naturally into trinkets like Eye of Magtheridon, and will work out of the box with `SimStats` when computing hit% in the output.

Reply:
SimStats context: https://github.com/marisa-ashkandi/tbcsim/blob/main/src/commonMain/kotlin/sim/SimStats.kt#L215

Reply:
Nice I went with `DAMAGE` event so that works, as I mentioned above it's nice seeing the output in the ability log as well. There are a few other places which have `SPELL_CAST` remaining with other classes, but I can put up a fix PR later for those if no one else gets to em.